pynest/examples/hpc_benchmark.py

```diff
@@ -0,0 +1,426 @@
+
+    if params['record_spikes']:
+        local_neurons = list(get_local_nodes(E_neurons))
+
+        if len(local_neurons) < brunel_params['Nrec']:
+            nest.message(M_ERROR, 'build_network',
+                         '''Spikes can only be recorded from local neurons, but the number of local
+                         neurons is smaller than the number of neurons spikes should be recorded from.
+                         Aborting the simulation!''')
+            exit(1)
+
+        nest.message(M_INFO, 'build_network', 'Connecting spike detectors.')
+        nest.Connect(local_neurons[:brunel_params['Nrec']], E_detector)
+
+    # read out time used for building
+    BuildEdgeTime = time.time() - tic
+
+    logger.log(str(BuildEdgeTime) + ' # build_edge_time')
+    logger.log(str(memory_thisjob()) + ' # virt_mem_after_edges')
+
+
+def run_simulation():
+    # open log file
+    logger = Logger(params['log_file'])
+
+    nest.ResetKernel
+    nest.set_verbosity(M_INFO)
+
+    logger.log(str(memory_thisjob()) + ' # virt_mem_0')
+
+    build_network(logger)
+
+    tic = time.time()
+
+    nest.Simulate(params['presimtime'])
+
+    PreparationTime = time.time() - tic
+
+    logger.log(str(memory_thisjob()) + ' # virt_mem_after_presim')
+    logger.log(str(PreparationTime) + ' # presim_time')
+
+    tic = time.time()
+
+    nest.Simulate(params['simtime'])
+
+    SimCPUTime = time.time() - tic
+
+    logger.log(str(memory_thisjob()) + ' # virt_mem_after_sim')
+    logger.log(str(SimCPUTime) + ' # sim_time')
+
+    logger.log(str(compute_rate()) + ' # average rate')
+
+    logger.done()
+
+# ------------------------------------------------------------------------------------
+
+
+def compute_rate():
+    '''compute approximation of average firing rate based on number of
+    local nodes, number of local spikes and total time
+    '''
+
+    n_local_spikes = nest.GetKernelStatus('local_spike_counter')
+    n_local_neurons = nest.GetKernelStatus('network_size') / nest.GetKernelStatus('num_processes')
+    simtime = nest.GetKernelStatus('time')
+    return 1. * n_local_spikes / (n_local_neurons * simtime) * 1e3
+
+#  ------------------------------------------------------------------------------------
+
+
+def memory_thisjob():
+    '''wrapper to obtain current memory usage'''
+    nest.sr('memory_thisjob')
+    return nest.spp()
+
+#  ------------------------------------------------------------------------------------
+
+
+def lambertwm1(x):
+    nest.sr('{} LambertWm1'.format(x))
+    return nest.spp()
+
+#  ------------------------------------------------------------------------------------
+
+
+def get_local_nodes(nodes):
+
+    '''
+    Generator for efficient looping over local nodes.
+    '''
+
+    nvp = nest.GetKernelStatus('total_num_virtual_procs')  # step size
+
+    i = 0
+    while i < len(nodes):
+        if nest.GetStatus([nodes[i]], 'local')[0]:
+            yield nodes[i]
+            i += nvp
```

Comment:
This is only correct if we have nodes with proxies, isn't it? That definitely needs to be documented in the docstring.

Reply:
I think it should also work for nodes without proxies. The minimal gid separation is nvp, isn't it? And in case that node is not local, we just start incrementing by one until we find a local node. At least that was the logic I had in mind. Updated the docstring.

Reply:
I see now. You are right. This will only work for nodes with proxies. The minimal step size for nodes without proxies could be one.